psychopy/experiment/components/sound/__init__.py

```diff
@@ -211,33 +211,35 @@ def writeFrameCode(self, buff):
         """
         # Write start code
         self.writeParamUpdates(buff, 'set every frame')
+
+        # write code for starting
         indented = self.writeStartTestCode(buff)
         if indented:
             if self.params['syncScreenRefresh'].val:
                 code = ("%(name)s.play(when=win)  # sync with win flip\n") % self.params
             else:
                 code = "%(name)s.play()  # start the sound (it finishes automatically)\n" % self.params
-            buff.writeIndented(code)
-        buff.setIndentLevel(-indented, relative=True)
-
-        # Write stop code
-        indented = self.writeStopTestCode(buff)
-        if indented:
-            code = ("%(name)s.stop()\n")
             buff.writeIndentedLines(code % self.params)
-        # because of the 'if' statement of the time test
         buff.setIndentLevel(-indented, relative=True)
 
-        # Update status
+        # update status according to playback
```

Comment:
@TEParsons I see your new implementation using `self.writeActiveTestCode(buff)`, but looking at this more carefully, I still struggle to see proper functioning of this activate status setting.

Let's look at one concrete example of Builder generated script:
```
            # if sound_1 is stopping this frame...
            if sound_1.status == STARTED:
                # is it time to stop? (based on global clock, using actual start)
                if tThisFlipGlobal > sound_1.tStartRefresh + 2-frameTolerance:
                    # keep track of stop time/frame for later
                    sound_1.tStop = t  # not accounting for scr refresh
                    sound_1.tStopRefresh = tThisFlipGlobal  # on global time
                    sound_1.frameNStop = frameN  # exact frame index
                    # add timestamp to datafile
                    thisExp.timestampOnFlip(win, 'sound_1.stopped')
                    # update status
                    sound_1.status = FINISHED
                    sound_1.stop()
            
            # if sound_1 is active this frame...
            if sound_1.status == STARTED:
                # update params
                pass
                # if sound is finished but sound_1 has time left, finish it now
                if sound_1.isFinished:
                    sound_1.status = FINISHED
```

If the sound finishes mid-frame, then the `if sound_1.isFinished:` will set `.status` to `FINISHED`. This means the actual stopping frame code on this next frame loop will get skipped. While users who are queuing `sound_1.status` will promptly know when the sound finished, but they have now lost access to updated attributes like `.tStop`, `.tStopRefresh`, and timestamp in data file on when the **sound component**. 

So my suggestion is not to use `writeActiveTestCode()`. Instead, we could overwrite `writeStopTestCode()` to add an `or` statement into the stopping frame code so it reads something like
```
            # if sound_1 is stopping this frame...
            if sound_1.status == STARTED:
                # is it time to stop? (based on global clock, using actual start)
                if tThisFlipGlobal > sound_1.tStartRefresh + 2-frameTolerance **or sound_1.isFinished**:
                    # keep track of stop time/frame for later
                    sound_1.tStop = t  # not accounting for scr refresh
                    sound_1.tStopRefresh = tThisFlipGlobal  # on global time
                    sound_1.frameNStop = frameN  # exact frame index
                    # add timestamp to datafile
                    thisExp.timestampOnFlip(win, 'sound_1.stopped')
                    # update status
                    sound_1.status = FINISHED
                    sound_1.stop()
```

I personally find this to be the most intuitive and elegant way of doing this. More generally, I think it could be useful to expand `writeStopTestCode()` to handle the use of additional stopping criteria by iterating over `self.params['stopType']` if it is a list, and then cascade multiple stopping criteria with `or`. Alternatively, we could add an internal `_extraStopType` parameter to `self.params`. Then `writeStopTestCode()` and `writeStopTestCodeJS()` just needs to be changed a little to accommodate this internal parameter that will be set by special components like a `SoundComponent`.

I'll make a PR to your branch to show you what I was thinking - can't promise it has handled all the details of `self.params` correctly...



This might be beyond the scope of a BF, so maybe overwriting `writeStopTestCode()` and `writeFrameCodeJS()`

Reply:
I like the idea of `writeStart/Stop/ActiveTestCode` functions accepting an additional condition - I think that's an elegant way around special cases like Sound where we need to query more than just `.status`. I don't think it needs to be anything as complicated as an internal parameter or differing behaviour for `stopType` - as the Component knows whether or not it needs it, we can just have it as an optional argument to those functions, called by the Components which need it.